libraries/common/test.effekt

```diff
@@ -56,10 +69,240 @@ def assertEqual[A](obtained: A, expected: A) { equals: (A, A) => Bool } { show:
+  try body() with Generator[Int] {
+    def generate() = resume {
+      while (true) {
+        do emit(randomInt(minInclusive, maxExclusive))
+      }
+    }
+
+    def shrink(v: Int) = resume {
+      var shrunk: Int = v/2
+      while (shrunk != 0 && shrunk != v) {
+        do emit(shrunk)
+        do emit(neg(shrunk))
+        shrunk = shrunk / 2 // in case of odd numbers, rounds to the smaller half (e.g. 5 -> round(2.5) = 2 )
+      }
+      do emit(0)
+    }
+  }
+
+def arbitraryBool { body: => Unit/Generator[Bool] }: Unit = 
+  try body() with Generator[Bool]{
+    def generate() = resume {
+      while (true) {
+        do emit(random() > 0.5)
+      }
+    }
+
+    def shrink(v: Bool) = resume {
+      var out = v
+      if(v){
+        out = false
+      }else {
+        out = true
+      }
+      do emit(out)
+    }
+  }
+
+def arbitraryDouble { body: => Unit/Generator[Double] }: Unit = 
+  try body() with Generator[Double]{
+    def generate() = resume {
+      val min = -1000.0 // TODO use double min value
+      val max = 1000.0  // TODO use double max value
+      while (true) {
+        do emit(min + (max - min) * random())
+      }
+    }
+
+    def shrink(v: Double) = resume {
+      var shrink: Double = v/2.0
+      while (shrink != 0.0 && shrink != v) {
+        do emit(shrink)
+        do emit(neg(shrink)) 
+        shrink = shrink / 2.0 // in case of odd numbers, rounds to the smaller half (e.g. 5 -> round(2.5) = 2 )
+      }
+      do emit(0.0)
+    }
+  }
+
+
+def arbitraryChar { body: => Unit / Generator[Char] }: Unit =
+  try body() with Generator[Char] {
+    def generate() = resume {
+      // ASCII printable characters range from 32 to 126
+      while (true) {
+        do emit(randomInt(32, 127).toChar) 
+      }
+    }
+
+    def shrink(v: Char) = resume { <> }
+  }
+
+def arbitraryString { body: => Unit / Generator[String] }: Unit / Generator[Char] = {
+  try body() with Generator[String] {
+    def generate() = resume {
+      while (true) {
+        val string: String = collectString { 
+          with boundary;
+          with limit[Char](randomInt(1, 20))
+          do generate[Char]() 
+        }
+        do emit(string)
+      }
+    }
+
+    def shrink(v: String) = resume { <> }
+  }
+}
+
+def arbitraryString(len: Int) { body: => Unit / Generator[String] }: Unit / Generator[Char] = {
+  try body() with Generator[String] {
+    def generate() = resume {
+      while (true) {
+        val string: String = collectString { 
+          with boundary;
+          with limit[Char](len)
+          do generate[Char]() 
+        }
+        do emit(string)
+      }
+    }
+
+    def shrink(v: String) = resume { <> }
+  }
+}
+
+def arbitraryList[T] { body: => Unit / Generator[List[T]]} : Unit / Generator[T] = {
+  try body() with Generator[List[T]]{ 
+    def generate() = resume {
+      while (true) {
+        val list = collectList[T]{ 
+          with boundary;
+          with limit[T](randomInt(1, 10))
```

Comment:
see my notes for the string: does this ever generate an empty list? isn't the limit too small?